README
1 README
********

The GNUstep CoreNetwork Library is a library of network-oriented
stream C objects.  For example, it includes types for HTTP or FTP
connections, support for SOCKS proxy servers and so on.

It aims to implement Apple's CFNetwork framework.

1.1 Initial reading
===================

The files `INSTALL' or `GNUstep-HOWTO' (from the web site) gives
instructions for installing the library.

1.2 License
===========

The GNUstep libraries and library resources are covered under the GNU
Lesser General Public License.  This means you can use these libraries in
any program (even non-free programs). If you distribute the libraries along
with your program, you must make the improvements you have made to the
libraries freely available. You should read the COPYING.LIB file for
more information. All files in the 'Source', 'Headers' directories
are covered under the GNU LGPL.

GNUstep test programs, and other files are covered under the
GNU General Public License. This means if you make changes to these
programs, you cannot charge a fee, other than distribution fees, for others
to use the program. You should read the COPYING file for more information.
All files in the 'Test', directory is covered under the GPL.

1.3 How can you help?
=====================

* Give us feedback!  Tell us what you like; tell us what you think
could be better.

Please log bug reports on the GNUstep project page
`http://savannah.gnu.org/bugs/?group=gnustep' or send bug reports
to <bug-gnustep@gnu.org>.

Happy hacking!

   Copyright (C) 2014 Free Software Foundation

   Copying and distribution of this file, with or without modification,
   are permitted in any medium without royalty provided the copyright
   notice and this notice are preserved.


ChangeLog
2014-02-03 Lubos Dolezel <lubos@dolezel.info>
	* Source/CFSocketStream.c:  Initial support for pre-existing sockets



Source/GSPrivate.h
/* GSPrivate.h
   
   Copyright (C) 2014 Free Software Foundation, Inc.
   
   This file is part of GNUstep CoreNetwork library.
   
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; see the file COPYING.LIB.
   If not, see <http://www.gnu.org/licenses/> or write to the 
   Free Software Foundation, 51 Franklin Street, Fifth Floor, 
   Boston, MA 02110-1301, USA.
*/

#ifndef __GSPRIVATE_H__
#define __GSPRIVATE_H__

#include <CoreFoundation/CFBase.h>
#include <CoreFoundation/CFRuntime.h>
#include "CoreFoundation/CFLocale.h"
#include "CoreFoundation/CFString.h"
#include "CoreFoundation/CFStringEncodingExt.h"



#if (defined(__GNUC__) && (__GNUC__ >= 4)) || defined(__clang__)
# define GS_PRIVATE __attribute__ ((visibility ("hidden")))
#else
# define GS_PRIVATE
#endif

struct __CFConstantString
{
  CFRuntimeBase  _parent;
  void          *_contents;
  CFIndex        _count;
  CFHashCode     _hash;
  CFAllocatorRef _deallocator;
};

#if defined (_WIN32)
#define DLL_EXPORT __declspec(dllexport)
#else
#define DLL_EXPORT
#endif

#define CONST_STRING_DECL(var, str) \
  static struct __CFConstantString __ ## var ## __ = \
    { {0, 0, {1, 0, 0}}, (void*)str, sizeof(str) - 1, 0, NULL }; \
  DLL_EXPORT const CFStringRef var = (CFStringRef) & __ ## var ## __;

#endif /* __GSPRIVATE_H__ */


Headers/CFNetwork/CFNetwork.h
/* CoreFoundation.h
   
   Copyright (C) 2014 Free Software Foundation, Inc.
   
   Written by: Lubos Dolezel
   Date: February, 2014
   
   This file is part of CoreNetwork.
   
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.         See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; see the file COPYING.LIB.
   If not, see <http://www.gnu.org/licenses/> or write to the 
   Free Software Foundation, 51 Franklin Street, Fifth Floor, 
   Boston, MA 02110-1301, USA.
*/ 

#ifndef __CFNETWORK_CFNETWORK_H__
#define __CFNETWORK_CFNETWORK_H__

#include <CoreFoundation/CFStream.h>
#include <CFNetwork/CFSocketStream.h>

#endif /* __CFNETWORK_CFNETWORK_H__ */



Headers/CFNetwork/CFSocketStream.h
/* CoreFoundation.h
   
   Copyright (C) 2014 Free Software Foundation, Inc.
   
   Written by: Lubos Dolezel
   Date: February, 2014
   
   This file is part of CoreNetwork.
   
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.         See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; see the file COPYING.LIB.
   If not, see <http://www.gnu.org/licenses/> or write to the 
   Free Software Foundation, 51 Franklin Street, Fifth Floor, 
   Boston, MA 02110-1301, USA.
*/ 

#ifndef __CFNETWORK_CFSOCKETSTREAM_H__
#define __CFNETWORK_CFSOCKETSTREAM_H__

#include <CoreFoundation/CFStream.h>

// TODO: many more declarations

#if OS_API_VERSION(MAC_OS_X_VERSION_10_2, GS_API_LATEST)
CF_EXPORT const CFStringRef kCFStreamPropertyShouldCloseNativeSocket;
#endif

#if OS_API_VERSION(MAC_OS_X_VERSION_10_3, GS_API_LATEST)
CF_EXPORT void
CFStreamCreatePairWithSocket (CFAllocatorRef alloc, CFSocketNativeHandle sock,
                              CFReadStreamRef *readStream,
                              CFWriteStreamRef *writeStream);
#endif

#endif /* __CFNETWORK_CFSOCKETSTREAM_H__ */



Source/CFSocketStream.c
/* CFSocketStream.c
   
   Copyright (C) 2014 Free Software Foundation, Inc.

   Written by: Lubos Dolezel
   Date: January, 2014
   
   This file is part of GNUstep CoreNetwork Library.
   
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.         See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; see the file COPYING.LIB.
   If not, see <http://www.gnu.org/licenses/> or write to the 
   Free Software Foundation, 51 Franklin Street, Fifth Floor, 
   Boston, MA 02110-1301, USA.
*/

#include "CFNetwork/CFSocketStream.h"
#include "GSPrivate.h"
#include <CoreFoundation/CFStreamPriv.h>
#include <CoreFoundation/CFNumber.h>

#ifndef _WIN32
#	include <sys/types.h>
#	include <sys/stat.h>
#	include <sys/socket.h>
#	include <arpa/inet.h>
#	include <fcntl.h>
#	include <unistd.h>
#	include <errno.h>
#	include <netdb.h>
#endif

CONST_STRING_DECL(kCFStreamPropertyShouldCloseNativeSocket, 
   "kCFStreamPropertyShouldCloseNativeSocket");

// Private functions defined by CoreBase
void
CFWriteStreamSetError(CFWriteStreamRef stream, int error);
void
CFReadStreamSetError(CFReadStreamRef stream, int error);

static void
CFWriteStreamSocketFinalize(CFWriteStreamRef s);
static void
CFWriteStreamSocketClose(CFWriteStreamRef s);
static CFTypeRef
CFWriteStreamSocketCopyProperty (CFWriteStreamRef s, CFStringRef propertyName);
static Boolean
CFWriteStreamSocketSetProperty (CFWriteStreamRef s, CFStringRef propertyName,
                          CFTypeRef propertyValue);
static CFIndex
CFWriteStreamSocketWrite (CFWriteStreamRef s, const UInt8 *buffer,
                    CFIndex bufferLength);
static Boolean
CFWriteStreamSocketOpen (CFWriteStreamRef s);

struct CFWriteStreamSocket
{
    struct __CFWriteStream    parent;
    int                       fd;
    int                       originalFd;
    Boolean                   closeOriginal;

    /* reference to related read stream for socket streams */
    struct CFReadStreamSocket *readStream;
};

static const struct CFWriteStreamImpl CFWriteStreamSocketImpl = {
    CFWriteStreamSocketClose,
    CFWriteStreamSocketFinalize,
    CFWriteStreamSocketOpen,
    CFWriteStreamSocketWrite,
    CFWriteStreamSocketCopyProperty,
    CFWriteStreamSocketSetProperty,
    NULL
};

static void
CFReadStreamSocketFinalize (CFReadStreamRef s);
static void
CFReadStreamSocketClose (CFReadStreamRef s);
static Boolean
CFReadStreamSocketOpen (CFReadStreamRef s);
static CFIndex
CFReadStreamSocketRead (CFReadStreamRef s, UInt8 *buffer, CFIndex bufferLength);
static Boolean
CFReadStreamSocketSetProperty (CFReadStreamRef s, CFStringRef propertyName,
                         CFTypeRef propertyValue);
static CFTypeRef
CFReadStreamSocketCopyProperty (CFReadStreamRef s, CFStringRef propertyName);

struct CFReadStreamSocket
{
    struct __CFReadStream      parent;
    int                        fd;
    int                        originalFd;
    Boolean                    closeOriginal;
    struct CFWriteStreamSocket *writeStream;
};

static const struct CFReadStreamImpl CFReadStreamSocketImpl = {
    CFReadStreamSocketClose,
    CFReadStreamSocketFinalize,
    CFReadStreamSocketOpen,
    CFReadStreamSocketRead,
    CFReadStreamSocketCopyProperty,
    CFReadStreamSocketSetProperty,
    NULL,
    NULL
};

#define CFWRITESTREAMSOCKET_SIZE (sizeof(struct CFWriteStreamSocket) - sizeof(CFRuntimeBase))
#define CFREADSTREAMSOCKET_SIZE (sizeof(struct CFReadStreamSocket) - sizeof(CFRuntimeBase))

void
CFStreamCreatePairWithSocket (CFAllocatorRef alloc, CFSocketNativeHandle sock,
                              CFReadStreamRef *readStream,
                              CFWriteStreamRef *writeStream)
{
  struct CFWriteStreamSocket* sfd;
  struct CFReadStreamSocket* rfd;

  *writeStream = (CFWriteStreamRef)
        _CFRuntimeCreateInstance (alloc, CFWriteStreamGetTypeID(),
                                  CFWRITESTREAMSOCKET_SIZE, 0);
  sfd = (struct CFWriteStreamSocket*) *writeStream;

  memcpy(&(*writeStream)->impl, &CFWriteStreamSocketImpl, sizeof(CFWriteStreamSocketImpl));
  sfd->fd = dup(sock);
  sfd->originalFd = sock;

  *readStream = (CFReadStreamRef)
        _CFRuntimeCreateInstance (alloc, CFReadStreamGetTypeID(),
                                  CFREADSTREAMSOCKET_SIZE, 0);
  rfd = (struct CFReadStreamSocket*) *readStream;

  memcpy(&(*readStream)->impl, &CFReadStreamSocketImpl, sizeof(CFReadStreamSocketImpl));
  rfd->fd = dup(sock);
  rfd->originalFd = sock;
}

static void
CFWriteStreamSocketClose(CFWriteStreamRef s)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;
  if (stream->fd != -1)
    {
      close(stream->fd);
      stream->fd = -1;
    }
}

static void
CFWriteStreamSocketFinalize(CFWriteStreamRef s)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;

  if (stream->readStream != NULL)
    {
      stream->readStream->writeStream = NULL;
    }
}

static CFTypeRef
CFStreamGetSocketProperty(int fd, CFStringRef propertyName)
{
  char hbuf[NI_MAXHOST];
  struct sockaddr* sa = NULL;
  socklen_t len = 0;

  getpeername(fd, sa, &len);
  
  if (errno != ENOBUFS)
    return NULL;

  sa = (struct sockaddr*) malloc(len);

  if (getpeername(fd, sa, &len) == -1)
    {
      free(sa);

      return NULL;
    }

  if (CFEqual(propertyName, kCFStreamPropertySocketRemoteHostName))
    {
      if (getnameinfo(sa, len, hbuf, sizeof(hbuf), NULL, 0, 0) == -1)
        {
          free(sa);

          return NULL;
        }

      free(sa);
      return CFStringCreateWithCString(NULL, hbuf, kCFStringEncodingISOLatin1);
    }
  else
    {
      int port = 0;

      if (sa->sa_family == AF_INET)
        port = ntohs( ((struct sockaddr_in*) sa)->sin_port );
      else if (sa->sa_family == AF_INET6)
        port = ntohs( ((struct sockaddr_in6*) sa)->sin6_port );

      free(sa);
      return CFNumberCreate(NULL, kCFNumberIntType, &port);
    }
}

static CFTypeRef
CFWriteStreamSocketCopyProperty (CFWriteStreamRef s, CFStringRef propertyName)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;

  if (CFEqual(propertyName, kCFStreamPropertyFileCurrentOffset))
    {
      CFIndex zero = 0;
      return CFNumberCreate(NULL, kCFNumberCFIndexType, &zero);
    }
  else if (CFEqual(propertyName, kCFStreamPropertySocketNativeHandle))
    {
      return CFDataCreate(NULL, (UInt8*) &stream->fd, sizeof(stream->fd));
    }
  else if (CFEqual(propertyName, kCFStreamPropertySocketRemoteHostName)
        || CFEqual(propertyName, kCFStreamPropertySocketRemotePortNumber))
   {
     CFTypeRef rv = CFStreamGetSocketProperty(stream->fd, propertyName);

     if (rv == NULL)
       CFWriteStreamSetError(s, errno);

     return rv;
   }
  else if (CFEqual(propertyName, kCFStreamPropertyShouldCloseNativeSocket))
    return stream->closeOriginal ? kCFBooleanTrue : kCFBooleanFalse;

  CFWriteStreamSetError(s, EINVAL);
  return false;
}

static Boolean
CFWriteStreamSocketSetProperty (CFWriteStreamRef s, CFStringRef propertyName,
                          CFTypeRef propertyValue)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;

  if (CFEqual(propertyName, kCFStreamPropertyShouldCloseNativeSocket))
    {
      stream->closeOriginal = propertyValue == kCFBooleanTrue;
      if (stream->readStream != NULL)
        stream->readStream->closeOriginal = stream->closeOriginal;
      return true;
    }

  CFWriteStreamSetError(s, EINVAL);
  return false;
}

static CFIndex
CFWriteStreamSocketWrite (CFWriteStreamRef s, const UInt8 *buffer,
                    CFIndex bufferLength)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;
  int wr;

  wr = write(stream->fd, buffer, bufferLength);
  if (wr == -1)
    CFWriteStreamSetError(s, errno);

  return wr;
}

static Boolean
CFWriteStreamSocketOpen (CFWriteStreamRef s)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;

  if (stream->fd != -1)
    return true; /* nothing to do */

  return false;
}

static void
CFReadStreamSocketClose (CFReadStreamRef s)
{
  struct CFWriteStreamSocket* stream = (struct CFWriteStreamSocket*) s;

  if (stream->fd != -1)
    {
      close(stream->fd);
      stream->fd = -1;
    }

  if (stream->closeOriginal && stream->readStream != NULL
      && stream->readStream->fd == 1)
    {
      close(stream->originalFd);
      stream->originalFd = -1;
    }
}

static Boolean
CFReadStreamSocketOpen (CFReadStreamRef s)
{
  struct CFReadStreamSocket* stream = (struct CFReadStreamSocket*) s;

  if (stream->fd != -1)
    return true; /* nothing to do */

  return false;
}

static CFIndex
CFReadStreamSocketRead (CFReadStreamRef s, UInt8 *buffer, CFIndex bufferLength)
{
  int rd;
  struct CFReadStreamSocket* stream = (struct CFReadStreamSocket*) s;

  rd = read(stream->fd, buffer, bufferLength);

  if (rd == -1)
    CFReadStreamSetError(s, errno);

  return rd;
}

static Boolean
CFReadStreamSocketSetProperty (CFReadStreamRef s, CFStringRef propertyName,
                         CFTypeRef propertyValue)
{
  struct CFReadStreamSocket* stream = (struct CFReadStreamSocket*) s;

  if (CFEqual(propertyName, kCFStreamPropertyShouldCloseNativeSocket))
    {
      stream->closeOriginal = propertyValue == kCFBooleanTrue;
      if (stream->writeStream != NULL)
        stream->writeStream->closeOriginal = stream->closeOriginal;
      return true;
    }

  CFReadStreamSetError(s, EINVAL);
  return false;
}

static CFTypeRef
CFReadStreamSocketCopyProperty (CFReadStreamRef s, CFStringRef propertyName)
{
  struct CFReadStreamSocket* stream = (struct CFReadStreamSocket*) s;

  if (CFEqual(propertyName, kCFStreamPropertyFileCurrentOffset))
    {
      CFIndex zero = 0;
      return CFNumberCreate(NULL, kCFNumberCFIndexType, &zero);
    }
  else if (CFEqual(propertyName, kCFStreamPropertySocketNativeHandle))
    {
      return CFDataCreate(NULL, (UInt8*) &stream->fd, sizeof(stream->fd));
    }
  else if (CFEqual(propertyName, kCFStreamPropertySocketRemoteHostName)
        || CFEqual(propertyName, kCFStreamPropertySocketRemotePortNumber))
   {
     CFTypeRef rv = CFStreamGetSocketProperty(stream->fd, propertyName);

     if (rv == NULL)
       CFReadStreamSetError(s, errno);

     return rv;
   }
  else if (CFEqual(propertyName, kCFStreamPropertyShouldCloseNativeSocket))
    return stream->closeOriginal ? kCFBooleanTrue : kCFBooleanFalse;

  CFReadStreamSetError(s, EINVAL);
  return false;
}

static void
CFReadStreamSocketFinalize(CFReadStreamRef s)
{
  struct CFReadStreamSocket* stream = (struct CFReadStreamSocket*) s;

  if (stream->writeStream != NULL)
    {
      stream->writeStream->readStream = NULL;
    }
}

